upstairs/src/lib.rs

```diff
@@ -1639,8 +1639,8 @@ impl Buffer {
 
     /// Consume and layer buffer contents on top of this one
     ///
-    /// The `buffer` argument will be reset to zero size, but preserves its
-    /// allocations for reuse.
+    /// The `buffer` argument will be left in an unknown state, and should be
+    /// reset before reuse.
```

Comment:
Could `eat` consume the argument `Buffer` then? This makes me uncomfortable, as it depends on the caller remembering to do something that isn't checked by the compiler.

Reply:
I was thinking about that, but I think we want to return the buffer so that it can be reused.  If you look at [`Volume::read`](https://github.com/oxidecomputer/crucible/blob/90b734824a07de2fcf654813bb39bee52d94116e/upstairs/src/volume.rs#L656-L700), we reuse `read_buffer` repeatedly.

I also convinced myself that resetting versus not-resetting doesn't make a difference, because `eat` was resetting the buffer to **size 0** (meaning we'd have to reset it _again_ anyways):

- If `eat` resets the buffer to 0 items, we have to call `Buffer::reset` or `Buffer::reset_owned` before using it again.  Since we're resizing from 0, we have to `memset` every item in both cases.
- If `eat` doesn't reset the buffer, we _still_ have to call `Buffer::reset` or `Buffer::reset_owned` before using it again.  However, `reset_owned` is a little cheaper, because we can skip calling `memset` on bytes in the range `0..previous size`.

Reply:
Addressed in 9907977ed5bec952f03af0dc977efa8b84d9d52a